http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ClusterHelper.scala

```diff
@@ -126,4 +320,94 @@ class ClusterHelper(dbRef: DbReference,
     }
   }
 
+  /* Process the templated cluster init script and put all initialization files in the init bucket */
+  private def initializeBucketObjects(cluster: Cluster, initBucketName: GcsBucketName, serviceAccountKey: Option[ServiceAccountKey]): Stream[IO, Unit] = {
+    // Build a mapping of (name, value) pairs with which to apply templating logic to resources
+    val replacements = ClusterInitValues(cluster, initBucketName, serviceAccountKey, dataprocConfig, proxyConfig, clusterFilesConfig, clusterResourcesConfig, contentSecurityPolicy).toMap
+
+    // Raw files to upload to the bucket, no additional processing needed.
+    val rawFilesToUpload = Stream.emits(
+      Seq(
+        clusterFilesConfig.jupyterServerCrt,
+        clusterFilesConfig.jupyterServerKey,
+        clusterFilesConfig.jupyterRootCaPem
+      )
+    ).evalMap { f =>
+      TemplateHelper.readFile[IO](f, executionContext).map(GcsResource(f, _))
+    }
+
+    val rawResourcesToUpload = Stream.emits(
+      Seq(
+        clusterResourcesConfig.jupyterDockerCompose,
+        clusterResourcesConfig.rstudioDockerCompose,
+        clusterResourcesConfig.proxyDockerCompose,
+        clusterResourcesConfig.proxySiteConf,
+        clusterResourcesConfig.welderDockerCompose,
+        clusterResourcesConfig.initVmScript
+      )
+    ).evalMap { resource =>
+      TemplateHelper.readResource(resource, executionContext).map(GcsResource(resource, _))
+    }
+
+    val templatedResourcesToUpload = Stream.emits(
+      Seq(
+        clusterResourcesConfig.initActionsScript,
+        clusterResourcesConfig.jupyterNotebookConfigUri,
+        clusterResourcesConfig.jupyterNotebookFrontendConfigUri
+      )
+    ).evalMap { resource =>
+      TemplateHelper.templateResource(replacements, resource, executionContext).map(GcsResource(resource, _))
+    }
+
+    val privateKey = Stream.emit(
+      for {
+        k <- serviceAccountKey
+        d <- k.privateKeyData.decode
+      } yield GcsResource(GcsBlobName(ClusterInitValues.serviceAccountCredentialsFilename), d.getBytes(Charset.defaultCharset))
+    ).unNone
+
+    (rawFilesToUpload ++ rawResourcesToUpload ++ templatedResourcesToUpload ++ privateKey).flatMap { resource =>
+      bucketHelper.storeObject(initBucketName, resource.gcsBlobName, resource.content, "text/plain")
+    }
+  }
+
+  private def getClusterVPCSettings(projectLabels: Map[String, String]): Option[Either[VPCNetworkName, VPCSubnetName]] = {
```

Comment:
shall we create a sum type for `VPCNetworkName` and `VPCSubnetName` instead of using `Either`?

Reply:
That probably makes sense.. will look into it.